core.ts

```diff
@@ -0,0 +1,1041 @@
+const noop = () => {}
+const deprecated: {
+	[ index: string ]: () => void
+} = {}
+const addDeprecated = (name: string, message: string) => {
+	deprecated[name] = () => {
+		console.warn('pinejs-client deprecated:', message)
+		deprecated[name] = noop
+	}
+}
+addDeprecated(
+	'expandFilter',
+	'`$filter: a: b: ...` is deprecated, please use `$filter: a: $any: { $alias: "x", $expr: x: b: ... }` instead.'
+)
+
+function defaults <T>(a: T | undefined, b: T | undefined, z: T): T
+function defaults <T>(a: T | undefined, z: T): T
+function defaults <T>(...args: (T | undefined)[]): T | undefined {
+	for (const arg of args) {
+		if (arg != null) {
+			return arg
+		}
+	}
+	return
+}
+const map = <T, R>(arr: T[], fn: (value: T) => R): R[] => {
+	const results = []
+	for (const key of arr) {
+		results.push(fn(key))
+	}
+	return results
+}
+const mapObj = <T, R>(obj: {[index: string]: T}, fn: (value: T, key: string) => R): R[] => {
+	const results = []
+	for (const key in obj) {
+		if (obj.hasOwnProperty(key)) {
+			results.push(fn(obj[key], key))
+		}
+	}
+	return results
+}
+
+const requiredUtilMethods = ['isString', 'isNumber', 'isBoolean', 'isObject', 'isArray', 'isDate']
+const isUtil = (obj: any): obj is PinejsClientCoreFactory.Util => {
+	if (obj == null) {
+		return false
+	}
+	for (const method of requiredUtilMethods) {
+		if (obj[method] == null) {
+			return false
+		}
+	}
+	return true
+}
+const isPromiseRejector = (obj: any): obj is PinejsClientCoreFactory.PromiseRejector => {
+	return obj != null && obj.reject != null
+}
+const isValidOption = (key: string): key is keyof PinejsClientCoreFactory.ODataOptions => {
+	return key === '$filter' ||
+		key === '$expand' ||
+		key === '$orderby' ||
+		key === '$top' ||
+		key === '$skip' ||
+		key === '$select'
+}
+
+// Workaround the fact that `setInterval` returns a different type in nodejs vs browsers
+const setIntervalResult = false as true && setInterval(() => {}, 0)
```

Comment:
So this is a fun one where we need to type something as the return type of `setInterval` (because the return type is different in nodejs vs the browser) and basically what the `false as true` does is make sure the bit after the `&&` is never run in reality, but the type system thinks it will always run and so `setIntervalResult` will be typed with the return type of that function - it's nasty but it's just the way to do it atm:
https://schneidenbach.gitbooks.io/typescript-cookbook/get-the-return-type-of-a-function.html
https://stackoverflow.com/questions/36015691/obtaining-the-return-type-of-a-function

That said it looks like typescript 2.8 (not released yet) is adding a special `ReturnType<setInteval>` which should handle this :D